googet_check.go

```diff
@@ -0,0 +1,119 @@
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+// The check subcommand searches the repo for packages using the filter provided. The default
+// filter is an empty string and will return all packages.
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/google/googet/v2/client"
+	"github.com/google/googet/v2/googetdb"
+	"github.com/google/googet/v2/goolib"
+	"github.com/google/googet/v2/install"
+	"github.com/google/googet/v2/system"
+	"github.com/google/logger"
+	"github.com/google/subcommands"
+)
+
+type checkCmd struct {
+	sources string
+	dryRun  bool
+}
+
+func (*checkCmd) Name() string     { return "check" }
+func (*checkCmd) Synopsis() string { return "check and take over exsiting packages" }
+func (*checkCmd) Usage() string {
+	return fmt.Sprintf(`%s check [-sources repo1,repo2...] [-dry_run=true]`, filepath.Base(os.Args[0]))
+}
+
+func (cmd *checkCmd) SetFlags(f *flag.FlagSet) {
+	f.BoolVar(&cmd.dryRun, "dry_run", false, "Don't make any changes to the DB.")
+	f.StringVar(&cmd.sources, "sources", "", "comma separated list of sources, setting this overrides local .repo files")
+}
+
+func (cmd *checkCmd) Execute(ctx context.Context, f *flag.FlagSet, _ ...interface{}) subcommands.ExitStatus {
+	exitCode := subcommands.ExitFailure
+	cache := filepath.Join(rootDir, cacheDir)
+	var filteredState []goolib.RepoSpec
+	db, err := googetdb.NewDB(filepath.Join(rootDir, dbFile))
+	if err != nil {
+		logger.Fatal(err)
+	}
+	defer db.Close()
+	state, err := db.FetchPkgs()
+	var newPkgs client.GooGetState
+	downloader, err := client.NewDownloader(proxyServer)
+	if err != nil {
+		logger.Fatal(err)
+	}
+	repos, err := buildSources(cmd.sources)
+	if err != nil {
+		logger.Fatal(err)
+	}
+	rm := downloader.AvailableVersions(ctx, repos, cache, cacheLife)
+	unmanaged := make(map[string]string)
+	fmt.Println("Searching for unmanaged software...")
+	for r, repo := range rm {
+		fmt.Println(r)
+		for _, p := range repo.Packages {
+			match := false
+			for _, k := range state {
+				if p.PackageSpec.Name == k.PackageSpec.Name {
+					match = true
+					break
+				}
+			}
+			if match {
+				continue
+			}
+			filteredState = append(filteredState, p)
+			app, _ := system.AppAssociation(p.PackageSpec.Authors, "", p.PackageSpec.Name, filepath.Ext(p.PackageSpec.Install.Path))
+			if app != "" {
+				unmanaged[p.PackageSpec.Name] = app
+				if cmd.dryRun {
+					continue
+				}
+				pi := goolib.PackageInfo{
+					Name: p.PackageSpec.Name,
+					Arch: p.PackageSpec.Arch,
+					Ver:  p.PackageSpec.Version,
+				}
+				if err := install.FromRepo(ctx, pi, r, cache, rm, archs, &newPkgs, true, downloader); err != nil {
```

Comment:
Following this state variable down into FromRepo, it gets passed to installDeps, then to minInstalled, which appears to believe that it represents the complete installed state of the system. But here we're passing an initially empty thing that maybe picks up a few packages through the loop. Is that going to break the dependency management?

Reply:
Dependencies will still evaluate and install. Even if we don't have full state, we can still evaluate dependencies and install

Reply:
Added some logic to handle this